Lib/zipfile.py

```diff
@@ -1818,7 +1818,9 @@ def close(self):
                              "Close the writing handle before closing the zip.")
 
         try:
-            if self.mode in ('w', 'x', 'a') and self._didModify: # write ending records
+            fp_open = not self.fp.closed if isinstance(self.fp, io.IOBase) else True
```

Comment:
There is a potential race condition here where `self.fp` could be closed after this statement but before `self._write_end_record` below.